src/ctap/storage.rs

```diff
@@ -187,338 +98,312 @@ impl PersistentStore {
         check_cred_protect: bool,
     ) -> Result<Option<PublicKeyCredentialSource>, Ctap2StatusCode> {
-        let key = Key::Credential {
-            rp_id: Some(rp_id.into()),
-            credential_id: Some(credential_id.into()),
-            user_handle: None,
-        };
-        let entry = match self.store.find_one(&key) {
-            None => return Ok(None),
-            Some((_, entry)) => entry,
-        };
-        debug_assert_eq!(entry.tag, TAG_CREDENTIAL);
-        let result = deserialize_credential(entry.data);
-        debug_assert!(result.is_some());
-        let user_verification_required = result.as_ref().map_or(false, |cred| {
-            cred.cred_protect_policy == Some(CredentialProtectionPolicy::UserVerificationRequired)
+        let mut iter_result = Ok(());
+        let iter = self.iter_credentials(&mut iter_result)?;
+        // We don't check whether there is more than one matching credential to be able to exit
+        // early.
+        let result = iter.map(|(_, credential)| credential).find(|credential| {
+            credential.rp_id == rp_id && credential.credential_id == credential_id
         });
-        if check_cred_protect && user_verification_required {
-            Ok(None)
-        } else {
-            Ok(result)
+        iter_result?;
+        if let Some(cred) = &result {
+            let user_verification_required = cred.cred_protect_policy
+                == Some(CredentialProtectionPolicy::UserVerificationRequired);
+            if check_cred_protect && user_verification_required {
+                return Ok(None);
+            }
         }
+        Ok(result)
     }
 
+    /// Stores or updates a credential.
+    ///
+    /// If a credential with the same RP id and user handle already exists, it is replaced.
     pub fn store_credential(
         &mut self,
-        credential: PublicKeyCredentialSource,
+        new_credential: PublicKeyCredentialSource,
     ) -> Result<(), Ctap2StatusCode> {
-        let key = Key::Credential {
-            rp_id: Some(credential.rp_id.clone()),
-            credential_id: None,
-            user_handle: Some(credential.user_handle.clone()),
-        };
-        let old_entry = self.store.find_one(&key);
-        if old_entry.is_none() && self.count_credentials()? >= MAX_SUPPORTED_RESIDENTIAL_KEYS {
+        // Holds the key of the existing credential if this is an update.
+        let mut old_key = None;
+        // Holds the unordered list of used keys.
+        let mut keys = Vec::new();
+        let mut iter_result = Ok(());
+        let iter = self.iter_credentials(&mut iter_result)?;
+        for (key, credential) in iter {
+            keys.push(key);
+            if credential.rp_id == new_credential.rp_id
+                && credential.user_handle == new_credential.user_handle
+            {
+                if old_key.is_some() {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+                }
```

Comment:
This is the style where you don't break, but rather check?

Reply:
I'm not sure what you mean by "style". But I guess you ask about the difference between `store_credential` and `find_credential` error handling. Here I prefer to "check" because it's a mutating operation, so doing the wrong thing would result in loss of data. For find credential, it's less problematic.